src/data/matrix/basic.lean

```diff
@@ -303,58 +312,62 @@ variable [fintype m]
 def dot_product [has_mul α] [add_comm_monoid α] (v w : m → α) : α :=
 ∑ i, v i * w i
 
+/- The precedence of 72 comes immediately after ` • ` for `has_scalar.smul`,
+   so that `r₁ • a ⬝ᵥ r₂ • b` is parsed as `(r₁ • a) ⬝ᵥ (r₂ • b)` here. -/
+localized "infix  ` ⬝ᵥ `:72 := matrix.dot_product" in matrix
+
 lemma dot_product_assoc [fintype n] [non_unital_semiring α] (u : m → α) (w : n → α)
   (v : matrix m n α) :
-  dot_product (λ j, dot_product u (λ i, v i j)) w = dot_product u (λ i, dot_product (v i) w) :=
+  (λ j, u ⬝ᵥ (λ i, v i j)) ⬝ᵥ w = u ⬝ᵥ (λ i, (v i) ⬝ᵥ w) :=
 by simpa [dot_product, finset.mul_sum, finset.sum_mul, mul_assoc] using finset.sum_comm
 
 lemma dot_product_comm [comm_semiring α] (v w : m → α) :
-  dot_product v w = dot_product w v :=
+  v ⬝ᵥ w = w ⬝ᵥ v :=
 by simp_rw [dot_product, mul_comm]
 
 @[simp] lemma dot_product_punit [add_comm_monoid α] [has_mul α] (v w : punit → α) :
-  dot_product v w = v ⟨⟩ * w ⟨⟩ :=
+  v ⬝ᵥ w = v ⟨⟩ * w ⟨⟩ :=
 by simp [dot_product]
 
 section non_unital_non_assoc_semiring
 variables [non_unital_non_assoc_semiring α] (u v w : m → α)
 
-@[simp] lemma dot_product_zero : dot_product v 0 = 0 := by simp [dot_product]
+@[simp] lemma dot_product_zero : v ⬝ᵥ 0 = 0 := by simp [dot_product]
 
-@[simp] lemma dot_product_zero' : dot_product v (λ _, 0) = 0 := dot_product_zero v
+@[simp] lemma dot_product_zero' : v ⬝ᵥ (λ _, 0) = 0 := dot_product_zero v
 
-@[simp] lemma zero_dot_product : dot_product 0 v = 0 := by simp [dot_product]
+@[simp] lemma zero_dot_product : 0 ⬝ᵥ v = 0 := by simp [dot_product]
 
-@[simp] lemma zero_dot_product' : dot_product (λ _, (0 : α)) v = 0 := zero_dot_product v
+@[simp] lemma zero_dot_product' : (λ _, (0 : α)) ⬝ᵥ v = 0 := zero_dot_product v
 
-@[simp] lemma add_dot_product : dot_product (u + v) w = dot_product u w + dot_product v w :=
+@[simp] lemma add_dot_product : (u + v) ⬝ᵥ w = u ⬝ᵥ w + v ⬝ᵥ w :=
 by simp [dot_product, add_mul, finset.sum_add_distrib]
 
-@[simp] lemma dot_product_add : dot_product u (v + w) = dot_product u v + dot_product u w :=
+@[simp] lemma dot_product_add : u ⬝ᵥ (v + w) = u ⬝ᵥ v + u ⬝ᵥ w :=
 by simp [dot_product, mul_add, finset.sum_add_distrib]
 
 end non_unital_non_assoc_semiring
 
 section non_unital_non_assoc_semiring_decidable
 variables [decidable_eq m] [non_unital_non_assoc_semiring α] (u v w : m → α)
 
-@[simp] lemma diagonal_dot_product (i : m) : dot_product (diagonal v i) w = v i * w i :=
+@[simp] lemma diagonal_dot_product (i : m) : (diagonal v i) ⬝ᵥ w = v i * w i :=
```

Comment:
```suggestion
@[simp] lemma diagonal_dot_product (i : m) : diagonal v i ⬝ᵥ w = v i * w i :=
```
And a few more of these below. It's only the ones with `λ` that need to keep the parentheses.

Reply:
Yes, I noticed exactly these uses. However, it looked too weird without the parentheses, so I put them back.

Is it really idiomatic to not wrap applications of functions of arity > 1 into parentheses when an operator follows/precedes?

Reply:
Yes, I think this is pretty standard. A search for `\w+ \w+ \w+ \* \w+ \w+` finds 224 results for me, such as:

https://github.com/leanprover-community/mathlib/blob/0b9663036339f050000277c99fab463b2e610f58/src/algebra/free.lean#L374

(sorry, getting github links from vscode search results is tiresome so I only grabbed the first)

Reply:
Don't you find `map f hf x * map f hf y` harder to read than `(map f hf x) * (map f hf y)` with parentheses? I do!

Reply:
Let's ask the community!
https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Lean.2Fmathlib.20style.20and.20readability

Reply:
Only with my python or C hat on. With my lean or haskell hat on, that looks fine. You get used to it eventually!

Reply:
Are all redundant uses or parentheses bad? They often help my eyes to group things correctly before I start with understanding what I am reading.

Reply:
If knowledge of whether the parentheses are needed requires knowledge of the precedence of some weird mathlib-specific operator means, then it's probably sensible to keep them around. Precedence of function application should be obvious though, so that one shouldn't need parentheses.

An example that is better with parentheses is `mul_assoc : x * y * z = x * (y * z)`, which is not at all obvious without the parens on the LHS.

Reply:
Thanks for your patient explanation!